<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Promises</title>
</head>
<body>
    
    <script>

        //new Promise()
            //Promise é uma função construtora de promessas. Existem dois resultados possíveis de uma promessa, ela pode ser resolvida, com a execução do primeiro argumento, ou rejeitada se o segundo argumento for ativado.

            const promessa0 = new Promise(function(resolve, reject) {
                let condicao = true //sempre com uma condição que se usa a promessa
                if (condicao) {
                    resolve()
                } else {
                    reject()
                }
            })

            console.log(promessa0)// Promise {<resolved>: undefined}
//-----------------------------------------------------------------------------------------------------

        // Resolve()
            //Podemos passar um argumento na função resolve(), este será enviado junto com a resolução da Promise.

            const promessa1 = new Promise(function(resolve, reject) {
            let condicao = true;
            if(condicao) {
                resolve('Estou pronto!');
            } else {
                reject();
            }
            });

            console.log(promessa1); // Promise {<resolved>: "Estou pronto!"}
//-----------------------------------------------------------------------------------------------------

        // Reject()
            //Quando a condição de resolução da promise não é atingida, ativamos a função reject para rejeitar a mesma. Podemos indicar no console um erro, informando que a promise foi rejeitada.

            const promessa2 = new Promise(function(resolve, reject) {
            let condicao = false;
            if(condicao) {
                resolve('Estou pronto!');
            } else {
                reject(Error('Um erro ocorreu.'));
            }
            });

            console.log(promessa2); // Promise {<rejected>: Error:...}
//-----------------------------------------------------------------------------------------------------

                                    //prototype

        //then() - then é o que torna uma promise especial. (then vem do prototype de promise)
            //O poder das Promises está no método then() do seu protótipo. O Callback deste método só será ativado quando a promise for resolvida. O argumento do callback será o valor passado na função resolve.
            //obs: o then é como se fosse um eventlistener da promessa, ou seja, fica esperando ser resolvida para fazer algo.

            const promessa3 = new Promise((resolve, reject) => {
            let condicao = true;
            if(condicao) {
                resolve('Estou pronto!');
            } else {
                reject(Error('Um erro ocorreu.'));
            }
            });

            promessa3.then(resultado => console.log(resultado))// 'Estou pronto!'
/*
            mesma coisa que:
            promessa3.then(function(resolucao) {
                console.log(resolucao) // 'Estou pronto!'
            })
*/
//-----------------------------------------------------------------------------------------------------

        //Assíncrono
            //As promises não fazem sentido quando o código executado dentro da mesma é apenas código síncrono. O poder está na execução de código assíncrono que executará o resolve() ao final dele.

            const promessa4 = new Promise(function(resolve, reject) {
                let condicao = true
                if (condicao) {
                    setInterval(function(params) {
                        resolve('Resolvida')
                    },1000)
                } else {
                    reject(Error('Ocorreu um erro.'))
                }
            })

            promessa4.then(params => console.log(params))
//-----------------------------------------------------------------------------------------------------

        //then().then()
            //O método then() retorna outra Promise. Podemos colocar then() após then() e fazer um encadeamento de promessas. O valor do primeiro argumento de cada then, será o valor do retorno anterior.

            const promessa5 = new Promise(function(resolve, reject) {
                resolve('Etapa 1')
            })
            promessa5.then(function(params) {
                console.log(params) // Etapa 1
                return 'Etapa 2'
            }).then(function(params) {
                console.log(params) // Etapa 2
                return 'Etapa 3'
            }).then(function (r){ //posso chamar o params do que quiser que ele continuca reconhecendo. (observe: antes era params e agora r)
                return r + 4
            })
            .then(function (params) { 
                console.log(params) // Etapa 34
            })
            // Obs IMPORTANTE: repare que o ponto está na linha de baixo e mesmo assim, funciona. Isso acontece porque o ponto tenta reconhecer o ultimo objeto ignorando os espaços, mas se escrevesse algo entre o ponto e o objeto, ja não daria certo. Isso serve para deixar o código mais organizado.
/*
            //mesma coisa que:

            const promessa5 = new Promise((resolve, reject) => {
            resolve('Etapa 1');
            });

            promessa5.then(resolucao => {
            console.log(resolucao); // 'Etapa 1';
            return 'Etapa 2';
            })

            .then(resolucao => {
            console.log(resolucao) // 'Etapa 2';
            return 'Etapa 3';
            })

            .then(r => r + 4)
            
            .then(r => {
            console.log(r); // Etapa 34
            });
*/
//-----------------------------------------------------------------------------------------------------

        //catch()
            //O método catch(), do protótipo de Promises, adiciona um callback a promise que será ativado caso a mesma seja rejeitada.

            const promessa6 = new Promise(function(resolve, reject) {
                let condicao = false

                if (condicao) {
                    resolve('Estou pronto, promessa 6!')
                } else {
                    reject(Error('Ocorreu um erro, promessa 6.'))
                }
            })

            promessa6.then(function(params){
                console.log(params)
            })
            .catch(function (reject) {
                console.log(reject + 'Erro menó')
            })
/*
            mesma coisa que:

            promessa6.then(params => console.log(params))
            .catch(reject => console.log(reject + 'Erro menó'))
*/

            // da pra passar direto o erro no then, como segundo argumento: (ver abaixo)
//-----------------------------------------------------------------------------------------------------
        
        // then(resolve, reject)
            //Podemos passar a função que será ativada caso a promise seja rejeitada, direto no método then, como segundo argumento.

            const promessa7 = new Promise((resolve, reject) => {
            let condicao = false;
            if(condicao) {
                resolve('Estou pronto!');
            } else {
                reject(Error('Um erro ocorreu. catch como segundo argumento do then'));
            }
            });

            promessa7.then(resolucao => {
            console.log(resolucao);
            }, reject => { //reject é o params
            console.log(reject);
            });
/*
            mesma coisa que: 
            promessa7.then(function(params){
                console.log(params)
            }, function(catchReject){ //passar catch como segundo argumento no then
                console.log(catchReject)
            })
*/
//-----------------------------------------------------------------------------------------------------

        //finally()
            //finally() executará a função anônima assim que a promessa acabar. A diferença do finally é que ele será executado independente do resultado, se for resolvida ou rejeitada.

            const promessa8 = new Promise((resolve, reject) => {
            let condicao = true; //true ou false, sai 'acabou no console'
            if(condicao) {
                resolve('Estou pronto!');
            } else {
                reject(Error('Um erro ocorreu.'));
            }
            });

            promessa8.then(resolucao => {
            console.log(resolucao);
            }, reject => {
            console.log(reject);
            }).finally(() => { //não recebe params/valor (se passar valor, vai sair undefined)
            console.log('Acabou'); // 'Acabou'
            })
//-----------------------------------------------------------------------------------------------------
                                    // Métodos do objeto promise

        //Promise.all()
            // no .all() da pra passar uma array com diferentes promessas. E o all vai retornar uma promessa também que só será resolvida quando todas as promises dentro da array forem resolvidas ou quando uma for rejeitada.

            //exemplo: serve pra quando você tem um servidor que tem uma promessa pro usuário fazer um login, depois que faz o login, espera para carregar os dados. Então com o all, você pode esperar o usuário fazer o login, esperar os dados serem carregados e fazer algo depois disso.

            //como se fosse [login, carregamento-dados]

            const loginUser = new Promise((resolve, reject) =>{ //usando prototype para criar uma nova
                setTimeout(() => {
                    resolve('Usuário logado')
                },1000)
            })

            const loadDados = new Promise((resolve, reject) =>{
                setTimeout(() => {
                    resolve('Dados carregados')
                },1500)
            })

            const carregouTudo = Promise.all([loginUser, loadDados]) //usando método da promise

            console.log(carregouTudo) //dentro do valor dessa promise tem os dois itens (da array) e esses dois itens são o resultado das promises.

            // E já que carregouTudo é uma promise cujo o valor é o valor dos dois itens, da pra usar o then nela:
            
            carregouTudo.then((resultado) => {
                console.log(resultado) //retorna array - ['Usuário logado', 'Dados carregados']
            })
//-----------------------------------------------------------------------------------------------------

        //Promise.race()
            //Retornará uma nova promise assim que a primeira promise for resolvida ou rejeitada. Essa nova promise terá a resposta da primeira resolvida. Ou seja, a primeira que sair ele já retorna.

            //exemplo: é útil quando está puxando dados de um servidor e não sabe quanto tempo vai levar pro servidor carregar, então com o race, é dizer, o primeiro dado que carregar já mostra pro usuário na tela. obs: com o fecth que é um tipo de promise, da pra puxar dados de um servidor.

            const login = new Promise(resolve => {
            setTimeout(() => {
                resolve('Login Efetuado');
            }, 1000);
            });
            const dados = new Promise(resolve => {
            setTimeout(() => {
                resolve('Dados Carregados');
            }, 1500);
            });

            const carregouPrimeiro = Promise.race([login, dados]);

            carregouPrimeiro.then(resposta => {
            console.log(resposta); // Login Efetuado
            });
    </script>
</body>
</html>